core/json/src/main/scala/net/liftweb/json/Meta.scala

```diff
@@ -108,55 +109,80 @@ private[json] object Meta {
     }
 
     def toArg(name: String, genericType: Type, visited: Set[Type], context: Context): Arg = {
-      def mkContainer(t: Type, k: Kind, valueTypeIndex: Int, factory: Mapping => Mapping) =
+      def mkContainer(t: Type, k: Kind, valueTypeIndex: Int, factory: Mapping => Mapping) = {
         if (typeConstructor_?(t)) {
           val typeArgs = typeConstructors(t, k)(valueTypeIndex)
           factory(fieldMapping(typeArgs)._1)
-        } else factory(fieldMapping(typeParameters(t, k, context)(valueTypeIndex))._1)
+        } else {
+          factory(fieldMapping(typeParameters(t, k, context)(valueTypeIndex))._1)
+        }
+      }
+
+      def mkHeteroContainer(baseType: Type): Mapping = {
+        val hereroContainerTypes = baseType match {
+          case ptype: ParameterizedType =>
+            ptype.getActualTypeArguments().map {
+              case c: Class[_] =>
+                c
+              case p: ParameterizedType =>
+                p.getRawType.asInstanceOf[Class[_]]
+              case x =>
+                fail("do not know how to get type parameter from " + x)
+            }
```

Comment:
Do we just want a `MatchError` if `baseType` isn't a `ParameterizedType`, or do we want to `fail` with our own error message?